src/lib/coda_lib/coda_lib.ml

```diff
@@ -55,13 +62,70 @@ let peek_frontier frontier_broadcast_pipe =
          (Error.of_string
             "Cannot retrieve transition frontier now. Bootstrapping right now.")
 
-let snark_worker_key t = t.config.snark_worker_key
+let client_port t =
+  let {Kademlia.Node_addrs_and_ports.client_port; _} =
+    t.config.net_config.gossip_net_params.addrs_and_ports
+  in
+  client_port
 
 let propose_public_keys t =
   Consensus.Data.Local_state.current_proposers t.config.consensus_local_state
 
 let replace_propose_keypairs t kps = Agent.update t.propose_keypairs kps
 
+let create_snark_worker ~logger ?(shutdown_on_disconnect = true) client_port =
+  let%bind p =
+    let our_binary = Sys.executable_name in
+    Process.create_exn () ~prog:our_binary
+      ~args:
+        ( "internal" :: Snark_worker.Intf.command_name
+        :: Snark_worker.arguments
+             ~daemon_address:
+               (Host_and_port.create ~host:"127.0.0.1" ~port:client_port)
```

Comment:
This use of `127.0.0.1` should be safe because we keep the worker ports distinct even if they have different localhost IPs.